### ISSUE TITLE
"Send Email" as part of a Content Rule should fail gracefully if no content for specifiied variable

### ISSUE THREAD

**Issue by rileydog** [plone/Products.CMFPlone#1358]
When using the "send email" as part of a content rule, if one of the variables included in the email does not exist, it generates an error (below) and stops the process.  The content rule should be designed to gracefully skip any variables that don't contain a value in the specific content type.  Also, the variable itself should not be included in the email.
    By taking this approach, the emails can be more generic, containing more variables that won't apply to all cases. For example, including the variables for comments, if there are any, but skipping if not.   The email being more generic allows the same content rule to be used in more places.    
Below is the error trace for the missing ${name}   (Note: as a test, I included all the variables available.  the first time I ran the content rule, I got an error because no comment available.  I deleted that part of the Rule, ran it again, then got this message.
=================================
Traceback (innermost last):
  Module ZPublisher.Publish, line 138, in publish
  Module ZPublisher.mapply, line 77, in mapply
  Module ZPublisher.Publish, line 48, in call_object
  Module plone.z3cform.layout, line 66, in __call__
  Module plone.z3cform.layout, line 50, in update
  Module plone.dexterity.browser.add, line 130, in update
  Module plone.z3cform.fieldsets.extensible, line 59, in update
  Module plone.z3cform.patch, line 30, in GroupForm_update
  Module z3c.form.group, line 145, in update
  Module plone.app.z3cform.csrf, line 21, in execute
  Module z3c.form.action, line 98, in execute
  Module z3c.form.button, line 315, in __call__
  Module z3c.form.button, line 170, in __call__
  Module plone.dexterity.browser.add, line 105, in handleAdd
  Module z3c.form.form, line 250, in createAndAdd
  Module plone.dexterity.browser.add, line 80, in add
  Module plone.dexterity.utils, line 182, in addContentToContainer
  Module Products.BTreeFolder2.BTreeFolder2, line 461, in _setObject
  Module zope.event, line 31, in notify
  Module zope.component.event, line 24, in dispatch
  Module zope.component._api, line 136, in subscribers
  Module zope.component.registry, line 321, in subscribers
  Module zope.interface.adapter, line 585, in subscribers
  Module plone.app.contentrules.handlers, line 181, in added
  Module plone.app.contentrules.handlers, line 122, in execute
  Module plone.contentrules.engine.executor, line 38, in __call__
  Module plone.contentrules.rule.rule, line 49, in __call__
  Module plone.app.contentrules.actions.mail, line 138, in __call__
  Module plone.stringinterp.dollarReplace, line 58, in __call__
  Module string, line 205, in safe_substitute
  Module string, line 196, in convert
  Module plone.stringinterp.dollarReplace, line 36, in __getitem__
  Module plone.stringinterp.adapters, line 44, in __call__
  Module plone.stringinterp.adapters, line 295, in safe_call
  Module Products.CMFCore.WorkflowTool, line 263, in getInfoFor
WorkflowException: No workflow provides '${name}' information.

**Comment by Shriyanshagro** [comment]
Hey,
I understand this issue pretty well, and would like to work upon this.
But I'm not able to find the desired package which has to be debugged. Can someone help me with this?

**Comment by prakharjoshi** [comment]
@Shriyanshagro You can find the content rules to send emails here https://github.com/collective/collective.contentrules.mailtogroup . Also Welcome to plone community and hope you have good time here.

**Comment by Shriyanshagro** [comment]
@prakharjoshi I have looked through this repo, I found that rules are already written for mail variables at https://github.com/collective/collective.contentrules.mailtogroup/blob/master/collective/contentrules/mailtogroup/actions/mail.py. 
Am I missing something here??  because there are no other variables I could find out other than these.

**Comment by rileydog** [comment]
Hi @Shriyanshagro and @prakharjoshi  
   I'm the person who posted this issue.  I'm not a programmer, so I don't understand most of the code.  what the code needs is the ability to skip any variable that isn't part of the actual email message.  
   this would allow a more generic email to be prepared with variables that won't be used for every email.

**Comment by Shriyanshagro** [comment]
Yeah, I understand this well, but I'm not able to find more variables other than those available in the file I have mentioned in my previous comment. @prakharjoshi and @rileydog can you show some light upon it??

**Comment by davisagli** [comment]
Most variables for content rule emails are defined in this package: https://github.com/plone/plone.stringinterp

The content rule framework is in https://github.com/plone/plone.contentrules and https://github.com/plone/plone.app.contentrules

**Comment by Shriyanshagro** [comment]
@davisagli  I dug very deep into it and back-trace the error till "Module plone.app.contentrules.actions.mail, line 138, in __call__", but can't  find anything for 
"Module Products.CMFCore.WorkflowTool, line 263, in getInfoFor".  

Possible explanation for error are-
1.Workflow Rule exception in https://github.com/plone/plone.contentrules
2.Naming Convention Error for Var(format) in line 80 of https://github.com/plone/Products.CMFPlone/blob/master/Products/CMFPlone/DublinCore.py as ".format" is itself a  property in python.

Is there any way to know which particular information is not being fetched??

Also this back-tracking was a fun as it allow me to explore Plone Programming Sense better :)

**Comment by jensens** [comment]
Still relevant for Plone 6. Content rule email action crashes when a string interpolation variable has no value instead of skipping gracefully.
